### ISSUE TITLE
Insufficient floating point precision of object/player position.

### ISSUE THREAD

**Issue by TheTermos** [minetest/minetest#9346]
##### Minetest version
```
All
```

This is also related to https://github.com/minetest/minetest_game/issues/1667, without this modification introducing quality mobs is not going to be possible.

##### Summary
This is a reduced example of the engine handling object movement:
```
float pos = 0.0f;       // center of the world
float speed = 1.0f;     // 0.1 n/s somewhat slow
float dtime = 0.0123f;  // typical engine step duration
pos += speed*dtime;		// = 0.01230
```
So far so good, but the world is huge, let's see what happens if it's near the edge.
```
float pos = 300000.0f;  // 30000 * BS
float speed = 1.0f;     // 0.1 n/s somewhat slow
float dtime = 0.0123f;  // typical engine step duration
pos += speed*dtime;		// = 300000.00000
```

Yup, it won't't budge. 
Of course dtime varies from step to step, so the overall result is the object slowly glitching forward in spasms.

It's not only turtles and snails, velocity component can often go small, e.g. when moving in a direction close to one of world axes or accelerating, it makes object behavior inconsistent depending on world position and it also makes getting collision detection to work properly more difficult.

Currently used f32 (float) type precision is just insufficient for the job.

I hear there's no much performance benefit to using float over double, also to my knowledge LUA doesn't care about 32 bit floats. 
So, can anyone think of any reasons not to increase the precision? 

##### Steps to reproduce
I've been able to confirm it consistently by setting object's velocity to 0.1 near (0,0,0) and near (-30000,0,-30000).
Movement is nice and smooth near center, rough at the edges.

**Comment by SmallJoker** [comment]
Yes, the precision isn't that great near the world edges. This is somewhat a non-trivial issue/request.

**General issue**: The memory representation of IEEE floating point values _is not defined_. Luckily most platforms seem to use the same format, though.

**Some history**: Prior 5.0.0-dev, floats were sent as `(s32)(value * 1000.f)` integers. This resulted in various Lua errors when exceeding the value range.

**Current situation**: Plain `f32` floats are now sent over the network. There is compatibility code (kindly provided by p_gimeno) to ensure the format is exactly the way it _should be_.

As you might already foresee, there are various topics to consider when switching to `f64`:

 * May we drop the network compatibility again (see #7768)? If you want compatibility, there will be much overhead.
 * Is there reliable code as fallback to serialize `f64` values to a fixed memory representation?

**Comment by TheTermos** [comment]
> **Current situation**: Plain `f32` floats are now sent over the network. There is compatibility code (kindly provided by p_gimeno) to ensure the format is exactly the way it _should be_.

Ok, great, let's keep sending them f32s.
Even if serverside only, increasing precision will prevent precision errors from piling up, and the client will catch up because updates. 
There's still the problem of the client getting slightly different numbers, but it keeps desyncing even now, so there's nothing to lose here.

What say you?

> Yes, the precision isn't that great near the world edges. This is somewhat a non-trivial issue/request

I only chose the edge case because it's easy to demonstrate and reproduce, in reality precision errors have effects all over the place, not just edges, this is important.

**Comment by TheTermos** [comment]
@SmallJoker do you agree messing with network protocol isn't required to make things better? Can we lose the non trivial label?

**Comment by scrain777** [comment]
While this is certainly non-trivial, a possible path forward would be to transmit a pair of f32 over the network, a reference point (say the center of the current block, or something larger) and an offset from the reference point. You could use the same reference point for everything relevant to a player, and only update it when the player moves a substantial amount. Then, inside the client and server it is optional whether to keep the pair of f32 or combine them into a f64. Collision detection would work fine with f32 and relative locations, as long as all the collision boxes share a reference point.

In principle the reference point could be s16, etc, but I believe that FP ops are faster than integer or, worse, combined integer-FP ops on most hardware these days.

**Comment by jordan4ibanez** [comment]
Wasn't this fixed in the fast math problem?

**Comment by appgurueu** [comment]
> Wasn't this fixed in the fast math problem?

No.

**Comment by Oblomov** [comment]
I support @scrain777 motion to split the position information. Coming from my experience in scientific computing with particle methods, we've been able to massively improvethe accuracy of our numerical code without resorting to double-precision since by tracking particle information as an (integer) cell index + local position within the cell.

Doing this properly is definitely non-trivial, and it will require changes in the protocol, in the engine internals AND in the Lua API, but I believe hat it can be done “gracefully” with an appropriate transition period to allow adoption of the new APIs by the mods and games before the final switch is done internally.

The general idea would be to represent positions as a combination of the integer node position (`s16 nx, ny, nz;`, currently) referring to the node center, and a fractional number (`f32 fx, fy, fz;`) where _the magnitude of the float is kept between -0.5f and 0.5f_ (one inclusive, the other not). Obviously, the fraction part can be omitted (and assmed 0) for elements that do not need it (such as nodes).

The positions and delta between positions can be converted into a single floating-point value of simply as n+f, but for example the position delta between (n0, f0) and (n1, f1)  should be kept as (n1-n0, f1 - f0) until needed in floating-point representation. Similarly, increment should be split into an integer and fractional increment, e.g. with ni = floor(increment + 0.5f), nf = increment - ni, when the source is not already in split form.

Adopting this strategy would significantly improve the numerical robustness of the code, reduce hacky workarounds, and solve the associated issues (esp. near world edge, such as #11742). The downside is obviously more complex code, potentially slower code in Lua (I expect little to no impact on the C++ side), and the need for a math library to handle all the details in a way that is as transparent as possible to the user (both C++ and Lua side).

On the upside, said library could come in handy in the transition period, since it can be designed to work with the current representation to begin with, and switch to the uniformly accurate one when “the time is mature”.

If there is an interest towards this solution, I can help code the C++ side of things and give a hand to anyone willing to tackle the Lua side that may need assistance (I'm personally not as well-versed in Lua as I am in C++, so I'm sure more competent developers than me could have more success on that side).

**Comment by Oblomov** [comment]
An upgrade to “double precision everywhere” definitely has the advantage of simplicity, but there are downsides too.

The main downsides are increased storage requirement (16 more bits per component compared to split representation) and the non-uniformity of precision (the fractional part gets all 53 bits near the origin, but only 37 bits near the edge of the world, and even though this is still more than the 23 given by single-precision split representation, it comes at the cost of storage requirements that are higher than the number of bits gained).

In a prospect of extensibility where the integer node coordinates are allowed to grow to 32-bit (even in non-uniform sizes for the various directions), the split representation “promotes” transparently, whereas the switch to double-precision would suffer from a sudden loss of accuracy in the fractional part (53 to 21 at the domain edge —less than that given by the split representation). In this sense, the split representation is “more future proof” (a longer-term investment).

Performance-wise, the situation is unclear. Even on 64-bit architectures double-precision operations are still largely half-rate compared to single-precision ones, although this depends on some implementation details and compiler choices). Split representation also benefits from higher ILP (Instruction Level Parallelism) since the integer and fractional part can be computed independently in many cases, which can help reduce latency in instruction throughput.

**Comment by jordan4ibanez** [comment]
Why is the calculation based on it's physical position vs repositioning all calculations at a zeroed point in space as if the current object being calculated is the center of all existence in the universe and everything else is relative to this? You could convert the other's positions to a __float128 then crush it back down with a tail trim back down to 64 then 32 if really needed. That way your collision bounds are centered to this current position.

OR

You need to do the current mapblock * f32 within this mapblock to keep it within the tail end ultra high precision, so that the rendering engine, raycasting, and collision is precise on the object. So if the mapblock is at 16 times cubed and the object is at 31 x it would be a data pack position wise of (mapblock_x = 2 * 16.0) + 15.0 to reinflate this into a usable dynamic scalar value when rendering/blah blah blah when REAL float position is needed, but all other calculations can utilize this packing method to maintain ultra high precision. The map is u_16 so this can be bit shifted into a (1D * 3 x,y,z) 16 bit integer array to hold the mapblock positional data and a float 1D (1D * 3 x.0,y.0,z.0) array. This combined with the relative collision detection creates a situation where the calculation math is highly precise at the minor cost of complication. This function can be turned into a micro object API internal to the engine like  Pack get_object_pack_position(object) -> Pack or V3F get_object_position(object) -> V3F where Pack is an object with fields that are immutable from the packed position data above unless a specific call is executed to update, and where V3F is the 1D * 3 x,y,z float position in real space as the players are viewing them

**Comment by appgurueu** [comment]
> Why is the calculation based on it's physical position vs repositioning all calculations at a zeroed point in space as if the current object being calculated is the center of all existence in the universe and everything else is relative to this? You could convert the other's positions to a __float128 then crush it back down with a tail trim back down to 64 then 32 if really needed. That way your collision bounds are centered to this current position.

This sounds like a massive overcomplication for no good reason. Just use f64 and stop worrying about precision issues as we do in Lua. For rendering, what you describe is already being done (because there it *does* matter whether you use f64 or f32).

**Comment by jordan4ibanez** [comment]
> > Why is the calculation based on it's physical position vs repositioning all calculations at a zeroed point in space as if the current object being calculated is the center of all existence in the universe and everything else is relative to this? You could convert the other's positions to a __float128 then crush it back down with a tail trim back down to 64 then 32 if really needed. That way your collision bounds are centered to this current position.
> 
> This sounds like a massive overcomplication for no good reason. Just use f64 and stop worrying about precision issues as we do in Lua. For rendering, what you describe is already being done (because there it _does_ matter whether you use f64 or f32).

Sounds like a massive oversimplification of an extremely complex problem

**Comment by appgurueu** [comment]
> > This sounds like a massive overcomplication for no good reason. Just use f64 and stop worrying about precision issues as we do in Lua. For rendering, what you describe is already being done (because there it _does_ matter whether you use f64 or f32).
> 
> Sounds like a massive oversimplification of an extremely complex problem

~~That's programming for you.~~ Please elaborate: Where does this approach fall short? The massive advantage I see is that this is practically already compatible with the number representation used by Lua all the time, so there's no need to cast when interfacing with objects from Lua.

**Comment by jordan4ibanez** [comment]
> > > This sounds like a massive overcomplication for no good reason. Just use f64 and stop worrying about precision issues as we do in Lua. For rendering, what you describe is already being done (because there it _does_ matter whether you use f64 or f32).
> > 
> > 
> > Sounds like a massive oversimplification of an extremely complex problem
> 
> ~That's programming for you.~ Please elaborate: Where does this approach fall short? The massive advantage I see is that this is practically already compatible with the number representation used by Lua all the time, so there's no need to cast when interfacing with objects from Lua.

I just realized it doesn't matter what we say here because no matter how many words are spoken in this pr it is dead and the collision will still be poopoo in 5 years so I hope you have a nice day

**Comment by appgurueu** [comment]
> I just realized it doesn't matter what we say here because no matter how many words are spoken in this pr it is dead and the collision will still be poopoo in 5 years so I hope you have a nice day

While this is an issue and not a PR, I unfortunately have to agree with your conclusion. I'll have a nice day nevertheless though. Note that these comments ought to be marked as spam eventually.

**Comment by tristan-cottam** [comment]
I have no experience in the field, but floating point decimals, while fit for general use, seem suboptimal for coordinate storage when fixed around a constant point (as opposed to entities themselves, e.g. the player), assuming location-independant precision is expected. Large coordinates are inaccurate, and small ones excessively precise.

Fixed point decimals (implemented as integers) would obviously require a colossal overhaul of the engine, not to mention being a significant breaking change, but seem like an optimal solution long-term.

On a sidenote, this issue seems like it should be moved to minetest/irrlicht.

**Comment by appgurueu** [comment]
> I have no experience in the field, but floating point decimals, while fit for general use, seem suboptimal for coordinate storage when fixed around a constant point (as opposed to entities themselves, e.g. the player), assuming location-independant precision is expected. Large coordinates are inaccurate, and small ones excessively precise.

Agreed. From some basic testing it seems the inaccuracies might be as large as 1/1000  near the world border (e.g. `31000.001f` is rounded to `31000.002`).

> Fixed point decimals (implemented as integers) would obviously require a colossal overhaul of the engine, not to mention being a significant breaking change, but seem like an optimal solution long-term.

I'm fairly certain that doubles, despite being floating point and thus having theoretically undesirable properties, would be fully sufficient since they just have so damn many mantissa bytes.

> On a sidenote, this issue seems like it should be moved to minetest/irrlicht.

No, not at all. Floats are used *inside Minetest*. Collisions are implemented in Minetest. This also pertains to Irrlicht which uses floats for rendering, but Minetest already offsets everything relative to the current node pos when rendering (since you need a maximum of precision for rendering, especially for depth tests).

**Comment by Oblomov** [comment]
Fixed point decimals per se provide no benefit in terms of world extension. It's the split representation that does it, and you can have it even with floating-point (integer node coordinate + floating-point node-relative offset).

**Comment by kno10** [comment]
Memory-wise, integers may actually be a decent way to go.
Instead of using a 32 bit float (single precision), using 16 bits node coordinate (=+-32k) and 16 bits within-block position is likely better. Computations on these values can use double precision, whose 53 bit is enough to be lossless.
Since some 10 or 12 bits within each node likely would be enough, this would allow for larger worlds.

The key benefit of floating point - the *floating* precision - is not that useful for minetest type of games. Positions close to block 0 do not need to be of a higher precision than positions far away. Rather use fixed point math then.

**Comment by appgurueu** [comment]
I see no need to overcomplicate this. I'm fairly convinced that doubles are going to be the simplest way to resolve this issue. For objects the additional memory usage of a couple bytes for position / velocity / maybe acceleration is completely negligible in comparison to all the other bloat they carry. I would strongly discourage optimizing for memory usage at the cost of code quality and possibly performance here. Doubles are the most obvious choice.

**Comment by Oblomov** [comment]
@kno10: my idea was to switch to int + float32. Same storage space as doubles, more uniform precision across the domain, and avoids the complexity of converting between fixed-point and floating-point when doing computations (and back to fixed-point when the result is stored). The only fixups you need to do is add/subtract 1 (from the floating and integer parts) when going below/above 0.5f.

@appgureu it is true that in the MT case the 53 bits of the fp64 significand are probably sufficient since the integer part of the coordinates takes 16 bits (at most, in floating-point), leaving 37 bits for the fractional part (which is more than the 24 bits you'd have with fp32 fractional). However, you still get different physics near block (0, 0, 0) compared to near blocks when any of the coordinates are closer to 2^16.

**Comment by kno10** [comment]
@Oblomov for a game like this, you want equal precision on an interval - fixed point precision does exactly that, and 32 bit may be enough here (but some users would like to have much larger worlds).

"int + float" is 64 bits, so it is more costly and more complicated than a 32 bit integer; and seems actually to be the worst possible combination: you have 32 bits for the cell id, and a worst-case in-cell precision (when close to 1) of about 5.9604645E-8 (the largest float values below 1 are 0.99999994 and 0.9999999) - about 24 bits of within-cell resolution (which is fairly useless IMHO). You will not use half of the exponent range if you cap the floats at 1!

A 64 bit double has 52+1 bits of mantissa; assuming that we use 32 bit for the node index that still gives at least 20 bits of within-cell-resolution, at a much simpler solution than your proposal. However, if we assume that 16 bit of within-cell resolution are sufficient, then it will support 4x larger worlds (where your integer runs out of precision).
A trivial 32 bit solution would use 16 bit for the node and 16 bit for the within-cell position. These integers can trivially be converted to double-precision by multiplication with 1p-16 (C++ base 2 exponent) for computations, but many computations could simply be done with the unscaled integers.

As Lua uses doubles everywhere, the Lua API will work fine with either of these two approaches. One is natively fp64, and the 32bit int approach one has an efficient lossless conversion to double.

To make the int+float approach competitive, you would need to use more bits of the mantissa for the position. If we continue to assume that 16 bit within-node precision is desirable, then you'd want the integer part to count 2^8 nodes. This eventually becomes a "custom float" with 55 bit mantissa and 8 bit exponent (as opposed to 52 and 11 for double, or 63 for signed integer with a fixed exponent).

Honestly, don't reinvent your own float this way. If IO matters, use a 32 bit with a fixed exponent (possibly even using varint encoding in the wire protocol), 64 bit integer, or double precision floats for simplicity.

**Comment by Oblomov** [comment]
@kno10 which solution is better depends entirely on what the objective is. Leaving aside custom floats (lots of work to get right, expensive to work with natively), the options we have are:

fixed point 16.16 (int16 + int16)
fixed point 32.32 (int32 + int32)
mixed 32.32 (int32 + fp32)
floating point 64 (fp64).

What determines which solution is “competitive” depends on (1) storage requirements  (2) desired degree of accuracy (3) desired _uniformity_ of accuracy (4) performance cost.

The only one competitive in terms of storage requirement is fixed point 16.16, since everything else requires 64-bits of storage.

In terms of _uniformity_ of accuracy, fp64 is the _worst_, if we're allowing the full range up to 2^32 nodes in one direction, since in that case the precision on the fractional part drops to to 20(+1) bits for nodes near the edge of the domain, which is worse than what you get with the next solution up, mixed 32.32, that gives you 24(+1) bits of (floating-point) precision _regardless of reference node_.

Yes, mixed 32.32 has the downside of not using half of the range of the exponent of the fp part, but that's not really that much of a loss.

Fixed-point 32.32 _does_ use the full range, but unless you're going to implement all operations in fixed-point directly, you're going to have to pass through floating-point anyway, in which case you'd have to promote the fractional part to fp64 because fp32 won't keep the full precision. So as far as the C++  side or any GPU shader is concerned, the extra precision that fixed-point would give you comes at a larger computational cost than mixed 32.32.

Of course as far as Lua is concerned, anything other than fp64 is going to require extensive code changes if the full range of the precision of the other formats has to be retained uniformly across the domain, since the split representation (node + fraction) needs to be preserved, which would double the size of all spatial variables in Lua.

**Comment by kno10** [comment]
@Oblomov first of all, doing the computation with integers is not actually hard to implement; and you are not at all restricted to 32.32 - you can use any split with fixed points easily, e.g., 48.16 and allow even larger worlds.
Most computations are trivially the same, without much accuracy issues: (double)(a.x-b.x) will be lossy only if a and b are *very* far away from each other, at which point the accuracy loss likely does not matter for gameplay. For any computations on nearby nodes, the conversion to float *after* taking the difference to *some* reference point (which may simply be the primary current map chunk!) will be just fine. This likely also involves shaders and all of that - just work in a coordinate system relative to some nearby map chunk (worst case is that you have to reload it every 2^10 nodes traveled?)

double will be as uniform as int+float for a range of up to 2^28 nodes *if* you insist on having 24 bits of precision within each node as in your int+float solution. If one decides that 2^10 resolution is enough within each cell, the double solution could address 2^43 nodes at 2^10 resolution. Its still making bad use of the exponent bits though, because an int64 solution could do 2^54 nodes at 2^10 resolution obviously. And as mentioned before, you could make and int+float solution that achieve 2^46 at 2^10 resolution, but that gets quite messy. But who needs such large worlds anyway?

On CPUs, int64 and double *will* be much more efficient than your int+float combination. On a GPU with shaders, you only need relative values to some fix point such as the node base coordinate, and you will be fine.
